server/coordinator/procedure/ddl/createpartitiontable/create_partition_table.go

```diff
@@ -0,0 +1,302 @@
+			if err := p.fsm.Event(eventCreatePartitionTable, createPartitionTableRequest); err != nil {
+				p.updateStateWithLock(procedure.StateFailed)
+				return errors.WithMessage(err, "create partition table")
+			}
+		case stateCreatePartitionTable:
+			if err := p.persist(ctx); err != nil {
+				return errors.WithMessage(err, "persist create partition table procedure")
+			}
+			if err := p.fsm.Event(eventCreateSubTables, createPartitionTableRequest); err != nil {
+				p.updateStateWithLock(procedure.StateFailed)
+				return errors.WithMessage(err, "create data tables")
+			}
+		case stateCreateSubTables:
+			if err := p.persist(ctx); err != nil {
+				return errors.WithMessage(err, "persist create partition table procedure")
+			}
+			if err := p.fsm.Event(eventFinish, createPartitionTableRequest); err != nil {
+				p.updateStateWithLock(procedure.StateFailed)
+				return errors.WithMessage(err, "update table shard metadata")
+			}
+		case stateFinish:
+			// TODO: The state update sequence here is inconsistent with the previous one. Consider reconstructing the state update logic of the state machine.
+			p.updateStateWithLock(procedure.StateFinished)
+			if err := p.persist(ctx); err != nil {
+				return errors.WithMessage(err, "create partition table procedure persist")
+			}
+			return nil
+		}
+	}
+}
+
+func (p *Procedure) Cancel(_ context.Context) error {
+	p.updateStateWithLock(procedure.StateCancelled)
+	return nil
+}
+
+func (p *Procedure) State() procedure.State {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	return p.state
+}
+
+type callbackRequest struct {
+	ctx context.Context
+	p   *Procedure
+}
+
+// 1. Create partition table in target node.
+func createPartitionTableCallback(event *fsm.Event) {
+	req, err := procedure.GetRequestFromEvent[*callbackRequest](event)
+	if err != nil {
+		procedure.CancelEventWithLog(event, err, "get request from event")
+		return
+	}
+	params := req.p.params
+
+	createTableMetadaResult, err := params.ClusterMetadata.CreateTableMetadata(req.ctx, metadata.CreateTableMetadataRequest{
+		SchemaName:    params.SourceReq.GetSchemaName(),
+		TableName:     params.SourceReq.GetName(),
+		PartitionInfo: storage.PartitionInfo{Info: params.SourceReq.PartitionTableInfo.GetPartitionInfo()},
+	})
+	if err != nil {
+		procedure.CancelEventWithLog(event, err, "create table metadata")
+		return
+	}
+	req.p.createPartitionTableResult = createTableMetadaResult
+}
+
+// 2. Create data tables in target nodes.
+func createDataTablesCallback(event *fsm.Event) {
+	req, err := procedure.GetRequestFromEvent[*callbackRequest](event)
+	if err != nil {
+		procedure.CancelEventWithLog(event, err, "get request from event")
+		return
+	}
+	params := req.p.params
+
+	for i, subTableShard := range params.SubTablesShards {
+		createTableResult, err := ddl.CreateTableMetadata(req.ctx, params.ClusterMetadata, params.SourceReq.GetSchemaName(), params.SourceReq.GetPartitionTableInfo().SubTableNames[i], subTableShard.ShardInfo.ID, nil)
```

Comment:
Does the index `i` of the `params.SubTablesShards` work for `params.SourceReq.GetPartitionTableInfo().SubTableNames`?

Reply:
`SubTableShard` is the shard assigned by the `shardPicker` to create subtables. They are consistent with the number of subtables, but they are not bound to the mapping relationship. Therefore, it is no problem to use `i` directly here.

Reply:
Maybe we should ensure the `len(SubTableNames) == len(params).SubTableShards` here by explicit assert.